package de.lambda9.ready2race.backend.app.qrCodeApp.boundary

import de.lambda9.ready2race.backend.app.App
import de.lambda9.ready2race.backend.app.ServiceError
import de.lambda9.ready2race.backend.app.appUserWithQrCode.control.AppUserWithQrCodeRepo
import de.lambda9.ready2race.backend.app.auth.entity.Privilege
import de.lambda9.ready2race.backend.app.qrCodeApp.control.QrCodeRepo
import de.lambda9.ready2race.backend.app.qrCodeApp.control.toRecord
import de.lambda9.ready2race.backend.app.qrCodeApp.entity.QrCodeError
import de.lambda9.ready2race.backend.app.qrCodeApp.entity.QrCodeUpdateDto
import de.lambda9.ready2race.backend.app.qrCodeApp.entity.toPublic
import de.lambda9.ready2race.backend.calls.responses.ApiResponse
import de.lambda9.ready2race.backend.calls.responses.ApiResponse.Companion.noData
import de.lambda9.ready2race.backend.database.generated.tables.records.AppUserWithPrivilegesRecord
import de.lambda9.ready2race.backend.kio.onTrueFail
import de.lambda9.tailwind.core.KIO
import de.lambda9.tailwind.core.extensions.kio.onNullFail
import de.lambda9.tailwind.core.extensions.kio.orDie

object QrCodeAppService {

    fun loadQrCode(
        qrCodeId: String,
        isAnonymous: Boolean,
    ): App<ServiceError, ApiResponse> = KIO.comprehension {
        val userOrParticipant = !QrCodeRepo.getUserOrParticipantByQrCodeIdWithDetails(qrCodeId).orDie()
        //.onNullFail { QrCodeError.QrCodeNotFound }

        when {
            userOrParticipant == null -> KIO.ok(ApiResponse.NoData)
            isAnonymous -> KIO.ok(ApiResponse.Dto(userOrParticipant.toPublic()))
            else -> KIO.ok(ApiResponse.Dto(userOrParticipant))
        }
    }

    private fun isQrCodeInUse(qrCodeId: String): App<ServiceError, Unit> = KIO.comprehension {
        val userOrParticipant = !QrCodeRepo.getUserOrParticipantByQrCodeId(qrCodeId).orDie()
        !KIO.failOn(userOrParticipant != null) { QrCodeError.QrCodeAlreadyInUse }
        KIO.unit
    }

    fun updateQrCode(
        update: QrCodeUpdateDto,
        user: AppUserWithPrivilegesRecord,
    ): App<ServiceError, ApiResponse> = KIO.comprehension {
        val record = update.toRecord(user.id!!)
        !isQrCodeInUse(record.qrCodeId)

        !AppUserWithQrCodeRepo.exists(eventId = record.event, qrCode = record.qrCodeId).orDie()
            .onTrueFail { QrCodeError.QrCodeAlreadyInUse }

        when (update) {
            is QrCodeUpdateDto.QrCodeAppuserUpdate ->
                if (!QrCodeRepo.appUserExistsWithQrCodeForEvent(update.id, record.event).orDie()) {
                    !QrCodeRepo.updateAppUserQrCode(
                        appUserId = update.id,
                        eventId = record.event
                    ) {
                        qrCodeId = record.qrCodeId
                    }.orDie()
                } else {
                    !QrCodeRepo
                        .create(record)
                        .orDie()
                }
            is QrCodeUpdateDto.QrCodeParticipantUpdate ->
                if (!QrCodeRepo.participantExistsWithQrCodeForEvent(update.id, record.event).orDie()){
                    !QrCodeRepo.updateParticipantQrCode(
                        participantId = update.id,
                        eventId = record.event
                    ) {
                        qrCodeId = record.qrCodeId
                    }.orDie()
                } else {
                    !QrCodeRepo
                        .create(record)
                        .orDie()

                }
        }
        noData

    }

    fun deleteQrCode(
        qrCodeId: String,
    ): App<ServiceError, ApiResponse> = KIO.comprehension {
        !QrCodeRepo.delete(qrCodeId).orDie()

        ApiResponse.noData
    }

}
